### ISSUE TITLE
[DPOTrainer] Tokenizer calculation fail during Q+A concat

### ISSUE THREAD

**Issue by GokulNC** [huggingface/trl#2054]
### System Info

trl 0.10.1

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

I am performing DPO on an instruction tuned model.

Since TRL does not apply chat template by default, I preprocess the data to add prompt template to each `prompt`.

This is the `chat_template` from my tokenizer:
```
{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = ' [INST] ' + message['content'] | trim + ' [/INST] ' %}{% if message['role'] != 'user' %}{% set content = message['content'] | trim + eos_token %}{% endif %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}
```

For instance, if the prompt text is `Choose the right option.`, after apply prompt template, it becomes:
`<s> [INST] Choose the right option. [/INST] `
(Note the final space)

When tokenized, it looks like this: (length: 10)
`[1, 67476, 4101, 31558, 4773, 4809, 67937, 67476, 4102, 67476]`

Now, since DPOTrainer concats the prompt + response directly as seen here:
https://github.com/huggingface/trl/blob/e2966c8d99104b6dda4f239d19a19ba36a7a6edc/trl/trainer/dpo_trainer.py#L313

It results in a string like this:
`<s> [INST] Choose the right option. [/INST] D`

which when tokenized, results in: (length 10)
`[1, 67476, 4101, 31558, 4773, 4809, 67937, 67476, 4102, 4884]`

Now the issue is that, in the DPOTrainer code, it results in empty list for the following line when computing answer tokens:
https://github.com/huggingface/trl/blob/e2966c8d99104b6dda4f239d19a19ba36a7a6edc/trl/trainer/dpo_trainer.py#L316-L317

When I trained a model without noticing this issue, it resulted in returning no response. That's when I started debugging this issue, and dug deeper to find this.

Now note that, the solution to this problem is not merely `strip`ing the prompt before concatenating. Because depending on whatever was the last few characters in the prompt/question, when a new answer string is appended, it might result in a different token at the intersection (if such a token exists in the vocabulary).

### Expected behavior

The `answer_input_ids` in the above code snippet should never become empty.

Although I am not sure what the solution is, as a sanity check, the code can first verify if `prompt_tokenized["input_ids"]` is fully present in `full_tokenized["input_ids"]`

**Comment by qgallouedec** [comment]
I think the issue comes from the chat template. What if you remove the final space in it?

**Comment by kashif** [comment]
@GokulNC  note that we have a special helper to check if merged tokens at the point of intersection are present and we try to deal with them appropriately

**Comment by qgallouedec** [comment]
Should be fixed via #2209
